File: backend/glasses_test/supabase_uploader.py
"""
Supabase Uploader for Glasses Pipeline
Uploads activity events, thumbnails, and daily lifestyle aggregations.
"""

import httpx
import json
import os
from datetime import datetime, date, timezone
from pathlib import Path
from uuid import UUID
from dotenv import load_dotenv

load_dotenv(Path(__file__).parent.parent / ".env")

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_SERVICE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")


class SupabaseUploader:
    """Upload glasses pipeline results to Supabase."""

    def __init__(self):
        self.base_url = f"{SUPABASE_URL}/rest/v1"
        self.storage_url = f"{SUPABASE_URL}/storage/v1"
        self.headers = {
            "apikey": SUPABASE_SERVICE_KEY,
            "Authorization": f"Bearer {SUPABASE_SERVICE_KEY}",
            "Content-Type": "application/json",
            "Prefer": "return=representation",
        }
        self.client = httpx.Client(timeout=30.0)

    def upload_activity_event(
        self,
        user_id: str,
        event_type: str,
        structured_data: dict,
        confidence_score: float,
        source_device: str = "ai_glasses",
        edge_model_version: str = "test-v0.1",
        thumbnail_ref: str = None,
    ) -> dict:
        """Upload a single activity event to user_activity_event table."""
        payload = {
            "user_id": user_id,
            "event_type": event_type,
            "source_device": source_device,
            "confidence_score": confidence_score,
            "structured_data": structured_data,
            "edge_model_version": edge_model_version,
            "verified": False,
            "processing_status": "edge_only",
        }
        if thumbnail_ref:
            payload["thumbnail_ref"] = thumbnail_ref

        resp = self.client.post(
            f"{self.base_url}/user_activity_event",
            headers=self.headers,
            json=payload,
        )
        resp.raise_for_status()
        return resp.json()

    def upload_thumbnail(self, bucket: str, file_path: str, data: bytes) -> str:
        """Upload thumbnail image to Supabase Storage. Returns public URL."""
        upload_headers = {
            "apikey": SUPABASE_SERVICE_KEY,
            "Authorization": f"Bearer {SUPABASE_SERVICE_KEY}",
            "Content-Type": "image/webp",
        }
        resp = self.client.post(
            f"{self.storage_url}/object/{bucket}/{file_path}",
            headers=upload_headers,
            content=data,
        )
        if resp.status_code == 400 and "already exists" in resp.text:
            # File exists, use upsert
            resp = self.client.put(
                f"{self.storage_url}/object/{bucket}/{file_path}",
                headers=upload_headers,
                content=data,
            )
        resp.raise_for_status()
        return f"{SUPABASE_URL}/storage/v1/object/public/{bucket}/{file_path}"

    def update_daily_lifestyle(
        self,
        user_id: str,
        recorded_date: str,
        total_calories: int,
        protein_g: int,
        fat_g: int,
        carb_g: int,
        sodium_mg: int,
        meal_count: int,
        data_source: str = "ai_glasses",
    ) -> dict:
        """Upsert daily lifestyle aggregation."""
        payload = {
            "user_id": user_id,
            "recorded_date": recorded_date,
            "total_calories": total_calories,
            "protein_g": protein_g,
            "fat_g": fat_g,
            "carb_g": carb_g,
            "sodium_mg": sodium_mg,
            "meal_count": meal_count,
            "data_source": data_source,
        }
        resp = self.client.post(
            f"{self.base_url}/user_lifestyle",
            headers=self.headers,
            json=payload,
        )
        resp.raise_for_status()
        return resp.json()

    def get_user_events_today(self, user_id: str) -> list[dict]:
        """Get all activity events for a user today."""
        today = date.today().isoformat()
        resp = self.client.get(
            f"{self.base_url}/user_activity_event",
            headers=self.headers,
            params={
                "user_id": f"eq.{user_id}",
                "detected_at": f"gte.{today}T00:00:00",
                "order": "detected_at.desc",
            },
        )
        resp.raise_for_status()
        return resp.json()

    def get_daily_summary(self, user_id: str) -> dict:
        """Calculate daily nutrition summary from today's events."""
        events = self.get_user_events_today(user_id)
        meals = [e for e in events if e.get("event_type") == "meal"]

        total_cal = 0
        total_protein = 0
        total_fat = 0
        total_carb = 0
        total_sodium = 0

        for meal in meals:
            data = meal.get("structured_data", {})
            nutrients = data.get("nutrients", {})
            total_cal += nutrients.get("energy_kcal", 0) or 0
            total_protein += nutrients.get("protein_g", 0) or 0
            total_fat += nutrients.get("fat_g", 0) or 0
            total_carb += nutrients.get("carbohydrate_g", 0) or 0
            total_sodium += nutrients.get("sodium_mg", 0) or 0

        return {
            "date": date.today().isoformat(),
            "meal_count": len(meals),
            "total_events": len(events),
            "total_calories": round(total_cal),
            "total_protein_g": round(total_protein),
            "total_fat_g": round(total_fat),
            "total_carb_g": round(total_carb),
            "total_sodium_mg": round(total_sodium),
        }

    def close(self):
        self.client.close()

File: frontend/HealthGlassesApp/src/services/SupabaseUploader.ts
/**
 * Supabase Upload Service
 * Uploads glasses pipeline results to Supabase backend
 *
 * Tables written to:
 *   - user_activity_event (individual meal/drink/medication events)
 *   - user_lifestyle (daily aggregated nutrition)
 *   - user_multimodal (if audio transcript stored)
 *
 * Storage:
 *   - thumbnails/ bucket (WebP thumbnails for every event)
 *   - clips/ bucket (full photos only when confidence < 0.7)
 */

import { PipelineResult } from './MLPipeline';

const SUPABASE_URL = 'https://oqotdxlmdgjieukyzegj.supabase.co';
const SUPABASE_ANON_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im9xb3RkeGxtZGdqaWV1a3l6ZWdqIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzYwNTY4MDYsImV4cCI6MjA5MTYzMjgwNn0.p2X5MKiGAalnQ0trit7nP7obEQzIcvoG4Ubu1X_kQFw';

interface UploadResult {
  success: boolean;
  eventId?: number;
  error?: string;
}

interface DailySummary {
  date: string;
  mealCount: number;
  totalEvents: number;
  totalCalories: number;
  totalProteinG: number;
  totalFatG: number;
  totalCarbG: number;
  totalSodiumMg: number;
}

class SupabaseUploaderService {
  private headers: Record<string, string>;

  constructor() {
    this.headers = {
      'apikey': SUPABASE_ANON_KEY,
      'Authorization': `Bearer ${SUPABASE_ANON_KEY}`,
      'Content-Type': 'application/json',
      'Prefer': 'return=representation',
    };
  }

  /**
   * Upload a pipeline result as an activity event
   */
  async uploadEvent(
    userId: string,
    result: PipelineResult,
    sourceDevice: string = 'meta_rayban_gen2',
  ): Promise<UploadResult> {
    const payload = {
      user_id: userId,
      event_type: result.eventType,
      source_device: sourceDevice,
      confidence_score: result.confidenceScore,
      structured_data: {
        food_items: result.foodDetections.map(d => ({
          name: d.foodName,
          confidence: d.confidence,
          portion: d.portionSize,
          portion_multiplier: d.portionMultiplier,
        })),
        nutrients: result.totalNutrients,
        processing_time_ms: result.processingTimeMs,
        faces_blurred: result.facesBlurred,
        audio_transcript: result.audioTranscript,
      },
      edge_model_version: result.modelVersions.foodClassifier,
      verified: false,
      processing_status: result.needsCloudReview ? 'needs_cloud_review' : 'edge_only',
    };

    try {
      const resp = await fetch(`${SUPABASE_URL}/rest/v1/user_activity_event`, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(payload),
      });

      if (!resp.ok) {
        const errText = await resp.text();
        return { success: false, error: `HTTP ${resp.status}: ${errText}` };
      }

      const data = await resp.json();
      return { success: true, eventId: data[0]?.event_id };
    } catch (err) {
      return { success: false, error: String(err) };
    }
  }

  /**
   * Upload thumbnail to Supabase Storage
   */
  async uploadThumbnail(
    userId: string,
    eventId: number,
    thumbnailBase64: string,
  ): Promise<string | null> {
    const filePath = `${userId}/${eventId}_thumb.webp`;
    const binaryData = this.base64ToBytes(thumbnailBase64);

    try {
      const resp = await fetch(
        `${SUPABASE_URL}/storage/v1/object/thumbnails/${filePath}`,
        {
          method: 'POST',
          headers: {
            'apikey': SUPABASE_ANON_KEY,
            'Authorization': `Bearer ${SUPABASE_ANON_KEY}`,
            'Content-Type': 'image/webp',
          },
          body: binaryData,
        },
      );

      if (resp.ok) {
        return `${SUPABASE_URL}/storage/v1/object/public/thumbnails/${filePath}`;
      }
    } catch (err) {
      console.log('[Upload] Thumbnail upload failed:', err);
    }
    return null;
  }

  /**
   * Get daily nutrition summary for a user
   */
  async getDailySummary(userId: string): Promise<DailySummary> {
    const today = new Date().toISOString().split('T')[0];

    try {
      const resp = await fetch(
        `${SUPABASE_URL}/rest/v1/user_activity_event?user_id=eq.${userId}&detected_at=gte.${today}T00:00:00&order=detected_at.desc`,
        { headers: this.headers },
      );

      const events = await resp.json();
      const meals = events.filter((e: any) => e.event_type === 'meal');

      let totalCal = 0, totalProtein = 0, totalFat = 0, totalCarb = 0, totalSodium = 0;

      for (const meal of meals) {
        const nutrients = meal.structured_data?.nutrients || {};
        totalCal += nutrients.energyKcal || 0;
        totalProtein += nutrients.proteinG || 0;
        totalFat += nutrients.fatG || 0;
        totalCarb += nutrients.carbohydrateG || 0;
        totalSodium += nutrients.sodiumMg || 0;
      }

      return {
        date: today,
        mealCount: meals.length,
        totalEvents: events.length,
        totalCalories: Math.round(totalCal),
        totalProteinG: Math.round(totalProtein),
        totalFatG: Math.round(totalFat),
        totalCarbG: Math.round(totalCarb),
        totalSodiumMg: Math.round(totalSodium),
      };
    } catch (err) {
      console.log('[Upload] Failed to get daily summary:', err);
      return {
        date: today, mealCount: 0, totalEvents: 0,
        totalCalories: 0, totalProteinG: 0, totalFatG: 0,
        totalCarbG: 0, totalSodiumMg: 0,
      };
    }
  }

  /**
   * Submit user correction for an event (active learning)
   */
  async submitCorrection(
    eventId: number,
    correctedFood: string,
    correctedCalories: number,
  ): Promise<boolean> {
    try {
      const resp = await fetch(
        `${SUPABASE_URL}/rest/v1/user_activity_event?event_id=eq.${eventId}`,
        {
          method: 'PATCH',
          headers: this.headers,
          body: JSON.stringify({
            verified: true,
            processing_status: 'user_verified',
            correction_data: {
              corrected_food: correctedFood,
              corrected_calories: correctedCalories,
              corrected_at: new Date().toISOString(),
            },
          }),
        },
      );
      return resp.ok;
    } catch {
      return false;
    }
  }

  /**
   * Upload daily lifestyle aggregation
   */
  async uploadDailyLifestyle(userId: string, summary: DailySummary): Promise<boolean> {
    try {
      const resp = await fetch(`${SUPABASE_URL}/rest/v1/user_lifestyle`, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify({
          user_id: userId,
          recorded_date: summary.date,
          total_calories: summary.totalCalories,
          protein_g: summary.totalProteinG,
          fat_g: summary.totalFatG,
          carb_g: summary.totalCarbG,
          sodium_mg: summary.totalSodiumMg,
          meal_count: summary.mealCount,
          data_source: 'ai_glasses',
        }),
      });
      return resp.ok;
    } catch {
      return false;
    }
  }

  private base64ToBytes(base64: string): Uint8Array {
    const binaryString = atob(base64.replace(/^data:.*,/, ''));
    const bytes = new Uint8Array(binaryString.length);
    for (let i = 0; i < binaryString.length; i++) {
      bytes[i] = binaryString.charCodeAt(i);
    }
    return bytes;
  }
}

export const supabaseUploader = new SupabaseUploaderService();
